Declarative pipeline for the monorepo elastic/beats.git

v1v opened the pull request:
### What

Even though https://github.com/elastic/beats/blob/7.8/Jenkinsfile is not anymore something we use, it's a valid complex declarative pipeline with several stages that we were using in our production environment [here](https://beats-ci.elastic.co/job/Beats/job/beats/).

cc @kuisathaverat

v1v committed:
Declarative pipeline for the monorepo elastic/beats.git

v1v commented on data/valid/monorepo-with-shared-library-steps/Jenkinsfile:
```diff
@@ -0,0 +1,1391 @@
+#!/usr/bin/env groovy
```
This is not what we want to evaluate, so I'll remove this shebang

v1v replied:
```suggestion
```

v1v committed:
Update data/valid/monorepo-with-shared-library-steps/Jenkinsfile

rtyler commented:
Oof! This is definitely a challenging one, thanks for the contribution!

I was wondering how soon somebody would come along with a Declarative Pipeline that had Groovy Script hiding in the eaves, turns out, immediately :smiling_imp:

v1v commented:
:)

kuisathaverat commented:
@v1v out of my head!!!! 😄

rtyler committed:
Allow just about any junk to be thrown into the file before `pipeline`

rtyler committed:
Parse more complex nested stages, etc

This adds some more logic to the parser rather than the grammar to ensure that
the stage {} directive has the necessary children blocks

The errors are pretty sweet too:

    ❯ ./target/debug/jdp check data/invalid/no-steps-in-stage/Jenkinsfile

    data/invalid/no-steps-in-stage/Jenkinsfile
    ------------------------------------------
    0: pipeline {
    1:     agent any
    2:     stages {
    3:         stage('Build') {
      --------^

    Fail: A stage must have either steps{}, parallel{}, or nested stages {}

rtyler commented:
So this `Jenkinsfiles` can be parsed, but I still have work to do in order to have the `jdp` CLI point out when there's scary potentially invalid Groovy hiding outside the `pipeline {}`
